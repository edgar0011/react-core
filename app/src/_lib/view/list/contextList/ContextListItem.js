// @flow

import React, { PureComponent } from 'react'
import type { Node } from 'react'
import { clickHandler } from '../../utils'
import type { HeadlessRender } from '../../common/HeadlessComponentBase'

type ContextListItemProps = {
  item: any,
  activeId?: number | string,
  idPropName?: string,
  title?: string,
  description?: string,
  active?: boolean,
  onItemClickHandler: Function,
} & HeadlessRender

/**
 * @class
 * @name ContextListItem
 * @desc Context List item component.
 * @prop {string} item Context List item to display.
 * @prop {number|string} activeId Optional identifier of active item inside Context List.
 * @prop {string} idPropName ??????? TODO Podlipny doplni
 * @prop {string} title Optional item title text string.
 * @prop {string} description Optional item description text string.
 * @prop {string} active Optional unique Context List identifier.
 * @prop {function} onItemClickHandler delegating onClick of an item
 * @prop {function} render Render function must be specified unless RenderComponent is specified.
 * @prop {ComponentType} RenderComponent Render component must be specified unless render function is specified.
 */
class ContextListItem extends PureComponent<ContextListItemProps> {
  constructor(props: ContextListItemProps) {
    super(props)
  }

  keyDownHandler = (event: SyntheticEvent<EventTarget>) => {
    if (clickHandler(event)) {
      const { item, idPropName, onItemClickHandler } = this.props
      onItemClickHandler(item[idPropName])
    }
  }

  props: ContextListItemProps

  render() {
    const {
      item, idPropName, activeId, onItemClickHandler, render, RenderComponent, ...props
    }: ContextListItemProps = this.props
    const isActive = activeId === item[idPropName]

    let rendered: Node
    if (render) {
      rendered = render({ ...props, ...item, active: isActive })
    } else {
      rendered = RenderComponent ? (
        <RenderComponent
          {...props}
          {...item}
          active={isActive}
        />
      ) : null
    }

    return (
      <div tabIndex='0' role='menu' onClick={onItemClickHandler(item[idPropName])} onKeyUp={this.keyDownHandler}>
        {rendered}
      </div>
    )
  }
}

export default ContextListItem
