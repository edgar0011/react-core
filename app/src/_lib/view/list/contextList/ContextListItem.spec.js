// TODO: move to core-fe
import React from 'react'
import ContextListItem from './ContextListItem'

describe('ContextListItem', () => {
  it('should render ContextListItem', () => {
    const props = {
      item: {},
      onItemClickHandler: jest.fn()
    }
    global.shallowSnapshot(<ContextListItem {...props} />)
  })
})
