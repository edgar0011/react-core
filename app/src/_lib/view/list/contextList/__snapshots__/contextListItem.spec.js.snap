// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextListItem should render ContextListItem 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ContextListItem
    item={Object {}}
    onItemClickHandler={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "onClick": undefined,
      "onKeyUp": [Function],
      "role": "menu",
      "tabIndex": "0",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "onClick": undefined,
        "onKeyUp": [Function],
        "role": "menu",
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
